### ISSUE TITLE
Packaging burden by breaking out pywbem code as separate libraries

### ISSUE THREAD

**Issue by tasleson** [pywbem/pywbem#2434]
I understand the desire to make internal code functionality into a separate library, but it comes at a cost.  For distributions like fedora that package each python library as a separate package, it requires a maintainer to create a new package, get the package through the approval process, and then maintain/test it over time.

At the moment those that are involved in pywbem packaging on fedora are considering dropping/orphaning pywbem in fedora as the packaging burden has increased (unless we can find someone to take it and all the new dependencies over).  If this occurs a few other packages in fedora which utilize pywbem will get dropped/orphaned too.

Is anyone here willing to help get these new packages specific to pywbem packaged for fedora and other distributions?

Maybe pywbem could come up with some other alternative, like git submodule?  Keep this needed functionality together and separate for reuse?  I'm not familiar with submodules and how difficult that would be do integrated releases in github for example.

Suggestions/help welcomed

Thanks

Ref. https://bugzilla.redhat.com/show_bug.cgi?id=1868128

**Comment by andy-maier** [comment]
Hi Tony.

We are certainly willing to help on this one.

* nocasedict and nocaselist are new packages on Pypi. They started out in pywbem, but are independent packages for use by anyone. They are both used by the normal operation of pywbem. The options I see are: (1) They become new packages on Fedora or (2) We vendor them into pywbem. Let me know which way to proceed.
* custom-inherit is likely a mistake on our end, I need to investigate that. It probably should have been only a dependency for development of pywbem. I'll post an update on this, next week (I'm on vacation tomorrow until monday).
* yamlloader is the replacement for yamlordereddictloader, as you found out. pywbem used yamlordereddictloader only in its tests. The newer yamlloader package got an additional use in pywbem now (in pywbem/_recorder.py). That's why you don't find a dependency to yamlordereddictloader in the history of the requirements.txt file of pywbem. Given that it is the recommended replacement for yamlordereddictloader, would adding yamlloader as a package be an issue?

**Comment by tasleson** [comment]
Hi @andy-maier 

We would appreciate `nocasedict` and `nocaselist` included/vendored if possible, unless you or someone else wants to pick up packaging for distros.  As for yamlloader, I will reach out to the fedora yamlordereddictloader maintainer and see if there is plans to replace it with yamlloader, which I think makes sense (assuming it's a drop in replacement).

I'll look into `custom-inherit` some more, but isn't it only required for testing?

Thanks for responding and a willingness to work with us.

**Comment by tasleson** [comment]
> On nocasedict/nocaselist: How would we go about packaging it for distros? Providing a spec file in their repos? Or having an account on the fedora site and creating the files there?

For fedora package maintainer ship see: 
https://fedoraproject.org/wiki/Join_the_package_collection_maintainers and
https://fedoraproject.org/wiki/Category:Package_Maintainers .

It's not a small amount of work and that's the reason I raised this issue.  Package  maintainers have a repository provided that contains your written spec file for the package and access to the tooling to build the package, integrated CI etc.  I'm thinking that for nocasedict/nocaselist it could possibly be done as a single package with two sub packages, but I'm not totally sure.  I'm not a packaging guru and know about all the ways to make the packaging simpler.  I don't have sufficient knowledge for other distributions and what it takes to maintain a package to comment about them. 

> custom-inherit is also used in normal use; we use it to avoid duplicate maintenance on docstrings. These are docstrings on classes a user would use. We discussed this in our team and we could abandon its use and either just keep the docstrings manually in sync, or have the derived class method provide a link to the base class method. You can assume its use will go away.

If the value add of it exceeds the work involved in packaging it, then one could argue that it's better to keep and package it.

For right now we are reverting our introduction of the 1+ version of pywbem in fedora and falling back to the old version until we figure out how to move forward.  This buys us some time.

So at the moment we are looking at 4 new packages being introduced for fedora.  I'm not sure what it looks like for debian and other distros/platforms.  Maybe @bzed can share some insight for debian.

**Comment by KSchopmeyer** [comment]
It is clear that at least custom_inherit is a disctaction to this discussion since it really gains the user nothing and was a convenience to avoid duplicate documentation between base classes and subclasses. I suggest that we agree to take this one off the table immediatly and use either hyperlinks for the documentation or duplicate it.

**Comment by andy-maier** [comment]
The dependency to the cutom-inherit package has been remove. PR #2438 rolls the removal back into the upcoming 1.0.3.

We will look into the packaging topic.

**Comment by andy-maier** [comment]
@tasleson I have started the process to create a package on Fedora for nocaselist. See this thread: https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/2Z6UTCTBNSXVYAMV2HOHUGII6IGZD7Z5/

The PR in our local nocaselist repo with the spec file is https://github.com/pywbem/nocaselist/pull/41, and it already contains resolutions for the comments in the mini-review by Robert.

I think I'm ready to fill in the package request form.

**Comment by andy-maier** [comment]
I also have created a packaging request for nocasedict: https://bugzilla.redhat.com/show_bug.cgi?id=1881285

So far, no reaction from RH.

**Comment by tasleson** [comment]
> So far, no reaction from RH.

Fedora is a project funded by RH.  However, fedora has many contributors that are not affiliated with RH, please see: https://www.redhat.com/en/topics/linux/fedora-vs-red-hat-enterprise-linux.

Please give folks time.  If you don't hear something after a while, I would suggest sending an email out on devel@lists.fedoraproject.org.  I see that you send an introductory email out, which is good.

**Comment by andy-maier** [comment]
ok.

**Comment by KSchopmeyer** [comment]
Moved to version 1.3

**Comment by andy-maier** [comment]
@tasleson Hi Tony. GIven that nobody reacted on the offer to maintain these packages in Fedora, the next best solution seems to be to vendor these two packages in pywbem. I am working on a PR that does this.

Would this be a workable solution for Fedora?

**Comment by tasleson** [comment]
@andy-maier If you're simply including them in the pywbem source tree, it should be good.  However, I believe there is one other package that pywbem switched too that hasn't been packaged for fedora.  I'll re-visit that today.

**Comment by andy-maier** [comment]
I did not want to include the source of the vendored packages in the pywbem repo. Instead, the pywbem build process pulls them in, so the package files as distributed on Pypi will have them inside.

Can you please double check whether your RPM build process can deal with that?

**Comment by tasleson** [comment]
> @tasleson I did not want to include the source of the vendored packages in the pywbem repo. Instead, the pywbem build process pulls them in, so the package files as distributed on Pypi will have them inside. This is done using the "vendorize" package on Pypi, which is now a build dependency.
> 
> This means your RPM build process will need to do something similar. Can you please double check whether your RPM build process can deal with that?

Technically we can do a lot of things in a rpm build, but from a policy standpoint I believe this is not allowed, pulling in source to complete the needed requirements of a build.  I'll pose the question on the fedora development mailing list to double check, but I think the answer will be no.  I believe the requirement is everything is either in the base tarball or separate rpm package.

If the released tarball of pywbem included these dependencies then I believe we would be OK.  Is that a possibility?

**Comment by tasleson** [comment]
@andy-maier I CC'd you on a new email thread.  Two things of note.

* There was some feedback on the BZ that doesn't appear to have any follow-up, see: https://bugzilla.redhat.com/show_bug.cgi?id=1880953#c1
* You may want to send email to python-sig list as suggested by Kevin in email thread.

**Comment by andy-maier** [comment]
@tasleson Tony, the released tarball and the released wheel archive both on Pypi and on the release page of the github repo do contain the vendored packages, because they are pulled in by the pywbem build process that creates them. Only the pywbem repo does not contain them.

**Comment by andy-maier** [comment]
@tasleson Tony, what are the current rules for supporting Python versions in Fedora?
I know that RHEL will support Python 2.7 as well as 3.5 and higher for the time being.

**Comment by tasleson** [comment]
> @tasleson Tony, what are the current rules for supporting Python versions in Fedora?
> I know that RHEL will support Python 2.7 as well as 3.5 and higher for the time being.

Fedora has support for a number of different python versions.  I would focus on compatibility with default in F33/F34 which I believe is Python 3.9.  Pywbem won't be updated in RHEL7, and likely not updated in RHEL8.  RHEL8 supports Python 3.9[1], so 3.9 would be an OK base minimum safe version IMHO.  You may want to support older Python for other distributions.  Python2 can and should be removed IMHO.

[1] https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/8.4_release_notes/new-features#enhancement_dynamic-programming-languages-web-and-database-servers

**Comment by tasleson** [comment]
> @tasleson Tony, the released tarball and the released wheel archive both on Pypi and on the release page of the github repo do contain the vendored packages, because they are pulled in by the pywbem build process that creates them. Only the pywbem repo does not contain them.

Then I think we might be OK, except for the yamlloader dependency.  Why is an order preserving yaml loader needed?

**Comment by andy-maier** [comment]
I vendored the yamlloader package as well. It is also in the archives.

It is needed because the operation recorder has a function to generate testcases which uses the ordered dumper in the yamlloader package to preserve the order of items, e.g. instances in a result. The function tests use the ordered loader in the yamlloader package to execute them. Note that "test" is an optional extension in the package.

**Comment by mayou36** [comment]
Hi all, yamlloader maintainer here. I read through the thread but I may miss out a detail: why can't yamlloader just be used? Or packaged? Is there any problem?

I am glad to assist from my side

**Comment by andy-maier** [comment]
@mayou36  Jonas, as I understand it, for Fedora there is a package maintainer for the predecessor package of your yamlloader, who is not interested in packaging your yamlloader. See Tony's comment https://github.com/pywbem/pywbem/issues/2434#issuecomment-878334349. If yamlloader was packaged in Fedora, we could simply have it in our dependency list (which we did before we attempted to resolve this packaging issue here).

**Comment by andy-maier** [comment]
@tasleson Tony, on the question of Python support, I am involved in a the [zhmc Ansible collection](https://github.com/zhmcclient/zhmc-ansible-modules) as well, and in that context we came across [this statement](https://access.redhat.com/solutions/4455511) of RedHat about Python 2 support which states that Python 2.7 is supported in RHEL 8 until june 2024. That is the reason we will not drop Python 2.7 support in this project until then.

**Comment by andy-maier** [comment]
@tasleson Thanks for pointing me to the package requests for nocasedict and nocaselist. They both now have a sponsor and I have followed up by updating the spec files from the review comments. If this results in an approval, I'll go ahead with the packaging process for these two, and would remove them again from the vendoring approach.

**Comment by tasleson** [comment]
> @tasleson Tony, on the question of Python support, I am involved in a the [zhmc Ansible collection](https://github.com/zhmcclient/zhmc-ansible-modules) as well, and in that context we came across [this statement](https://access.redhat.com/solutions/4455511) of RedHat about Python 2 support which states that Python 2.7 is supported in RHEL 8 until june 2024. That is the reason we will not drop Python 2.7 support in this project until then.
> 
> My earlier statement about RHEL supporting Python 3.5 I think was incorrect. That was taken from [this statement](https://docs.ansible.com/ansible/latest/reference_appendices/python_3_support.html#python-3-support) which is actually a statement about Ansible and not about RHEL.
> 
> However, in the [article you linked above](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/8.4_release_notes/new-features#enhancement_dynamic-programming-languages-web-and-database-servers) it says: "Note that Red Hat will continue to provide support for Python 3.6 until the end of life of RHEL 8".
> 
> Bottom line, my understanding is that RHEL 8 supports Python 2.7 and Python 3.6 (and some higher Python 3 versions).

For RHEL8 we are packaging python3 package only.  As I mentioned in  a previous comment, I don't see us doing a re-base in RHEL8.  If/when a bug rises in the version we are shipping, I'll back port a fix.  This is why I made my suggestion about pywbem current development moving to 3.9.

**Comment by andy-maier** [comment]
Ok, I understand the Python 3.x argument now.

On the front of package ownership for nocaselist and nocasedict, things have moved a bit. I have a potential sponsor for me being a packager and have sent in my review test results.

**Comment by andy-maier** [comment]
After such a long time with this issue open, I have to say that I have no time to pursue the packaging of the nocaselist and nocasedict dependencies in Fedora. I'm sorry for that, but that's how it is right now.

On the question whether we should vendor in the dependencies (nocasedict, nocaselist, yamlloader): Karl and I discussed that and we don't like that approach at all. We'll drop PR #2737 which was the attempt to vendor them in.

I'm afraid to say that at this point, pywbem in Fedora needs to stay below 1.0.0.

Closing this issue.
